### PR TITLE
feat: build binaries and Docker images in CI

### LINKED ISSUE #24773: Build, package, and sign binary builds, deb/rpms, and Docker for Releases
```[tasklist]
### Tasks
- [x] Build binaries for Windows, Mac, and Linux
- [x] Build release binaries for Windows, Mac, and Linux only on release
- [x] Have CI run the same on all platforms
- [x] Package builds into deb/rpms
- [x] Sign binary builds
- [x] Upload binary builds to s3
- [x] Create a docker build pipeline for release in the official images repo
```

### PR DESCRIPTION
**Pull request by mgattozzi**
feat: build binaries and Docker images in CI 

For releases we need to have Docker images and binary images available for the
user to actually run influxdb3. These CI changes will build the binaries on a
release tag and the Docker image as well, test, sign, and publish them and make
them available for download.

Co-Authored-By: Brandon Pfeifer <bpfeifer@influxdata.com>

Closes #24773

### CODE REVIEW

**Comment by mgattozzi** [comment]
Okay this is open for review. Before merging I will need to set the CI to only run only on release tags. If you want to test the builds yourself let me know internally on slack and I can get you access to the docker builds and the binary builds.

**Comment by mgattozzi** [comment]
I'm having some issues with the defaults I've set for docker via env vars (something about permission denied for the fs) that I don't get just running the binary itself which is a docker issue with how I'm running it I think. Otherwise I think we're in a great spot.

File: .circleci/packages/influxdb3/control/post-uninstall
Comment (hiltontj) [inline .circleci/packages/influxdb3/control/post-uninstall:1]: I am assuming this file just needs to be here to prevent something from breaking, despite it not executing anything.

File: .circleci/scripts/package-validation/validate
Comment (hiltontj) [inline .circleci/scripts/package-validation/validate:116]: Do we need to keep these commented lines?

**Review by hiltontj — APPROVED** [review]
Thanks @mgattozzi - this is a huge lift. I only had a couple questions/comments, but otherwise LGTM. Would probably be good to get @bnpfeife's approval as well.

File: .circleci/packages/influxdb3/control/post-uninstall
Comment (mgattozzi) [reply .circleci/packages/influxdb3/control/post-uninstall:1]: I might be able to remove it, but I'm not sure. @bnpfeife would have a better idea

File: .circleci/scripts/package-validation/validate
Comment (mgattozzi) [reply .circleci/scripts/package-validation/validate:116]: We could remove it and add them in later if we end up building out a systemd service

File: .circleci/packages/influxdb3/control/post-uninstall
Comment (bnpfeife) [reply .circleci/packages/influxdb3/control/post-uninstall:1]: Yup, you should be able to remove it. If that breaks anything, it's an issue with `ci-packager` that I will need to take a look at.

**Review by hiltontj — APPROVED** [review]
🚢

**Comment by mgattozzi** [comment]
With @hiltontj's changes to where `heappy` is located I'll need to make a few here in order to get the builds to work on windows again but it shouldn't be that hard. @bnpfeife if you think that the changes otherwise are good then I think we can merge once I fix the conflicts and turn on the release filters so CI isn't building artifacts all the time.

File: .circleci/packages/config.yaml
Comment (jdstrand) [inline .circleci/packages/config.yaml:7]: This default value is difficult to work with. Consider:
```
$ dpkg --compare-versions 3.0.0 lt 3.0.1 && echo yes  # good
yes

$ dpkg --compare-versions 3.0.0 lt 3.x-0123abcd && echo yes  # maybe not good?
yes

$ dpkg --compare-versions 3.x-gfed9876 lt 3.x-0123abcd && echo yes  # but what if 3.x-0123abcd was created before 3.x-gfed9876?
[1]
```

Part of the reason why I'm commenting on this is that we are uploading to `influxdb/releases/influxdb3`, but these aren't releases, they're snapshots. Package managers are going to have trouble dealing with these versions. I would advocate that the `influxdb/releases/influxdb3` only contain things that are considered a release and use something else for snapshots or nightlies.

Fyi, `telegraf` has `telegraf/nightlies/telegraf_nightly_amd64.deb`, which gets overwritten every night. Maybe that isn't what you want right now though (ie, perhaps it is useful to have different snapshots). 

As a strawperson, I suggest that:
1. someone creates `influxdb/snapshots/influxdb3` and move everything that was uploaded to `influxdb/releases/influxdb3` there. New snapshots go to `influxdb/snapshots/influxdb3` and use `3.0.0+snapshot-{{env.CIRCLE_SHA1[:8]}}` as the default.
2. if one of the current uploads was an mvp, tag it with `3.0.0-0.alpha.0.0` and put in `influxdb/releases/influxdb3`. For important internal pre-alpha releases, you can tag with `3.0.0-0.alpha.0.1`, `3.0.0-0.alpha.0.2`, etc. When official alpha 1 is released, tag with `3.0.0-0.alpha.1`, then use `3.0.0-0.alpha.2` for 2nd official alpha, and so on. Continue with `3.0.0-0.beta.1`, `3.0.0-0.rc.1` as desired. This versioning scheme works with both semver and packaging. Ie: `3.0.0-0.alpha.0.0` < `3.0.0-0.alpha.0.1` < `3.0.0-0.alpha.0.2` < `3.0.0-0.alpha.1` < `3.0.0-0.alpha.2` < `3.0.0-0.beta.1` < `3.0.0-0.rc.1`. All of these tagged releases go to `influxdb/releases/influxdb3`
3. when 3.0.0 is released, tag as `3.0.0`. semver is happy since `3.0.0-0.rc.1` < `3.0.0` and packaging is updated to use `3.0.0-1` such that `3.0.0-0.rc.1` < `3.0.0-1`

In the above, we don't worry if `3.0.0+snapshot-{{env.CIRCLE_SHA1[:8]}}` is higher or lower in packaging since people will be installing those manually.

**Review by jdstrand — CHANGES_REQUESTED** [review]

File: .circleci/packages/config.yaml
Comment (bnpfeife) [reply .circleci/packages/config.yaml:7]: This is _definitely_ a good point. I may go ahead and update these for 1.X OSS/Enterprise and 2.X OSS. As you alluded to, the default version schemes weren't intended to be consumed by the general public. So I had just written them to make `apt` and `yum` happy. However, I definitely can foresee a future where someone makes these available to the general public (not here, but in 1.X and 2.X land).

File: .circleci/config.yml
Comment (bnpfeife) [inline .circleci/config.yml:423]: Since we're publishing "snapshot" packages, let's change the destination depending on the release type:

```diff
   publish-packages:
     docker:
       - image: cimg/python:3.6
+    parameters:
+      # "destination" should be one of:
+      #   - releases
+      #   - nightlies
+      #   - snapshots
+      destination:
+        - type: string
      steps:
        - attach_workspace:
           at: /tmp/workspace
        - aws-s3/sync:
           arguments:             --acl public-read
           aws-region:            RELEASE_AWS_REGION
           aws-access-key-id:     RELEASE_AWS_ACCESS_KEY_ID
           aws-secret-access-key: RELEASE_AWS_SECRET_ACCESS_KEY
           from:                  /tmp/workspace/artifacts
-          to:                    s3://dl.influxdata.com/influxdb/releases
+          to:                    s3://dl.influxdata.com/influxdb/<< parameters.destination >>
```

The workflows will need to be modified to pass `destination` along to the jobs.

**Review by bnpfeife — CHANGES_REQUESTED** [review]

**Comment by mgattozzi** [comment]
I'll still need to address all of these. I did get the branch rebased to latest main and conflicts fixed. I'll get to these early next week when I'm back from the solar eclipse viewing.

**Comment by mgattozzi** [comment]
@jdstrand and @bnpfeife I made the requested changes. I also added the sed filtering you mentioned on slack @jdstrand to make using sha256sum --check work easier.

File: .circleci/config.yml
Comment (jdstrand) [inline .circleci/config.yml:423]: `cimg/python:3.6` here and elsewhere appears to be EOL based on https://circleci.com/developer/images/image/cimg/python

File: .circleci/config.yml
Comment (jdstrand) [inline .circleci/config.yml:428]: The `snapshots` bucket needs additional setup I think. From https://app.circleci.com/pipelines/github/influxdata/influxdb/38817/workflows/05b5eb9c-ef95-4d29-9465-c7dcd7a8a91a/jobs/362952:
```
upload failed: ../../../tmp/workspace/artifacts/influxdb3_3.0.0+snapshot-77fe062b-1_amd64.deb.sha256 to s3://dl.influxdata.com/influxdb/snapshots/influxdb3_3.0.0+snapshot-77fe062b-1_amd64.deb.sha256 An error occurred (AccessDenied) when calling the PutObject operation: Access Denied
```

File: .circleci/packages/config.yaml
Comment (jdstrand) [inline .circleci/packages/config.yaml:7]: This should be `^v3.0.0-0.alpha.[0-9]+.[0-9]+` (the `-0` is important for package managers).

File: .circleci/packages/config.yaml
Comment (jdstrand) [inline .circleci/packages/config.yaml:10]: You're using `v3...` above and `3...` here. Update to `v3.0.0+snapshot-{{env.CIRCLE_SHA1[:8]}}` here?

That said, I'd just assume remove `v` from the other two instead as it adds a complication when building debs and rpms, but perhaps it is make something else easier that I'm unaware of. Cc @bnpfeife

File: .circleci/packages/influxdb3/control/post-install
Comment (jdstrand) [inline .circleci/packages/influxdb3/control/post-install:13]: In `pre-install` you define:
```
USER=influxdb3
GROUP=influxdb3
```

It would be nice to do that here.

File: .circleci/packages/influxdb3/control/post-install
Comment (jdstrand) [inline .circleci/packages/influxdb3/control/post-install:11]: AIUI, this is saying that on Debian-based systems, unconditionally chown and chmod the LOG_DIR and DATA_DIR. Does this happen on upgrades or just new installs? Why is this needed if post-install is already creating the directories correctly. This feels like something cargo-culted in from other packaging that was meant to fix a permissions problem for the other package that doesn't apply here, but maybe I'm missing something?

**Review by jdstrand — CHANGES_REQUESTED** [review]
Thanks for these changes! See inline.

File: .circleci/packages/influxdb3/control/post-install
Comment (mgattozzi) [reply .circleci/packages/influxdb3/control/post-install:11]: This is a straight copy from oss 2.x so maybe @bnpfeife would know more?

File: Dockerfile
Comment (mgattozzi) [reply Dockerfile:64]: I can't remember, I think I wasn't sure or I was copying what we did in the other OSS bits. Happy to make it /var/lib.

File: .circleci/config.yml
Comment (mgattozzi) [reply .circleci/config.yml:428]: @bnpfeife could you help with that?

**Comment by mgattozzi** [comment]
There are still some things that need addressing but am waiting on @bnpfeife's feedback to implement them correctly.

File: .circleci/packages/config.yaml
Comment (bnpfeife) [reply .circleci/packages/config.yaml:10]: When `release.match` matches against `$CIRCLE_TAG`, `release.value` is supplied as the package version. Otherwise, `default.version` is supplied as the package version. In retrospect, the tool could probably consider an array of possibilities instead of just `release` and `default`. However, I think that _just_ `release` and `default` will suffice for our current needs.

The intent of `release.match` is to prevent building packages with versions incompatible with apt/deb and yum/rpm. Unfortunately, our release tags have included the `v` prefix (even since before I started). So, without wanting to cause trouble, I just made the tooling work around it. In most cases, `match` should include the `^v`, but `value` should not. `{{ env.CIRCLE_TAG[1:] }}` instructs the templating engine to strip the first character from `$CIRCLE_TAG`.

With all that being said, I think this is the correct specification:
```yaml
version:
  release:
    match: '^v[0-9]+.[0-9]+.[0-9]+'
    value: '{{env.CIRCLE_TAG[1:]}}'
  default:
    value: '3.0.0+snapshot-{{env.CIRCLE_SHA1[:8]}}'
```

Both release and alpha tags should retain everything except the `v`. In all other circumstances, `3.0.0+snapshot-{{env.CIRCLE_SHA1[:8]}}` will be used instead.

File: .circleci/config.yml
Comment (bnpfeife) [reply .circleci/config.yml:428]: Hey, I've created an issue with SRE: https://github.com/influxdata/sre/issues/3145

**Review by bnpfeife — CHANGES_REQUESTED** [review]

File: .circleci/config.yml
Comment (bnpfeife) [reply .circleci/config.yml:428]: All right, uploads into `dl.influxdb.com/influxdb/snapshots` are working now!

File: .circleci/packages/config.yaml
Comment (mgattozzi) [reply .circleci/packages/config.yaml:10]: So this will also work if say we do a tag of `v3.0.0-0.alpha.0.2`?

File: .circleci/config.yml
Comment (jdstrand) [reply .circleci/config.yml:428]: @bnpfeife - thanks for setting this up! I see that there are a lot of influxdb3 packages in `influxdb/releases` still. Would you be able to move those over to `influxdb/snapshots`?

**Review by jdstrand — COMMENTED** [review]
Thanks for all the changes! I have one open question for @bnpfeife, so adding this as a comment now.

File: .circleci/packages/config.yaml
Comment (jdstrand) [reply .circleci/packages/config.yaml:6]: > However, like @mgattozzi said, we want to support versions like 3.0.0-0.alpha.0.1, 3.0.0-0.alpha.1, 3.0.0-0.beta.1 and 3.0.0-0.rc.1 in addition to 3.0.0 (ie, release) and 3.0.0+snapshot-deadbeef (ie, default), but it's not clear (to me!) how those would fit in to what is there now (just default andrelease). 

I think this may 'just work', since `match: '^v[0-9]+.[0-9]+.[0-9]+'` should match all of those and so just end up with the tag, so long as the tag prefixes with `v`. Eg: `3.0.0-0.alpha.0.1`, `3.0.0-0.alpha.1`, `v3.0.0-0.beta.1`, `3.0.0-0.rc.1`, `v3.0.0`.

File: .circleci/config.yml
Comment (bnpfeife) [reply .circleci/config.yml:428]: All right, these have been moved!

File: .circleci/packages/config.yaml
Comment (jdstrand) [reply .circleci/packages/config.yaml:6]: We could be fancier and do:
```
match: '^v[0-9]+.[0-9]+.[0-9]+(-[0-9]+\.(alpha|beta|rc)\.[0-9]+(\.[0-9]+)?)?$'
```

Eg:
```
$ for i in v3.0.0-0.alpha.0.1 v3.0.0-0.alpha.1 v3.0.0-0.beta.1 v3.0.0-0.rc.1 v3.0.0 good-before-bad-after v3.0.0.alpha.0.1 v3.0.0.alpha.1 3.0.0-0.beta.1 v3.0.0-0.beta v3.0.0rc ; do echo "$i" | grep -E '^v[0-9]+.[0-9]+.[0-9]+(-[0-9]+\.(alpha|beta|rc)\.[0-9]+(\.[0-9]+)?)?$' ; done
v3.0.0-0.alpha.0.1
v3.0.0-0.alpha.1
v3.0.0-0.beta.1
v3.0.0-0.rc.1
v3.0.0
```
This would force us to conform to the version if we want a tag to end up in `releases/` that is an alpha, beta, rc or release.

**Comment by jdstrand** [comment]
> someone creates influxdb/snapshots/influxdb3 and move everything that was uploaded to influxdb/releases/influxdb3 there

Fyi, @bnpfeife did this just now. `aws s3api list-objects --no-sign-request --bucket dl.influxdata.com --prefix influxdb/releases` doesn't show any `influxdb3` packages now and they are all seen with `aws s3api list-objects --no-sign-request --bucket dl.influxdata.com --prefix influxdb/snapshots`.

Thanks!

File: .circleci/packages/config.yaml
Comment (bnpfeife) [reply .circleci/packages/config.yaml:6]: Okay, I manually supplied `CIRCLE_*` values to `ci-packager`. It looks like everything works as expected.

```sh
$ docker run -it --rm \
  --env CIRCLE_TAG=v3.0.0-0.alpha.1 \
  --env CIRCLE_SHA1="$(git rev-parse HEAD)" \
  --env CIRCLE_BRANCH="$(git rev-parse --abbrev-ref HEAD)" \
  -v "/tmp/workspace:/tmp/workspace" \
  -v "${PWD}:/project" \
  us-east1-docker.pkg.dev/influxdata-team-edge/ci-support/ci-packager:latest bash
  
$ cd /project && packager .circleci/packager/config.yaml
...

$ ls -l artifacts
total 232400
-rw-r--r-- 1 root root 46982921 Apr 11 20:32 influxdb3-3.0.0-0.alpha.1-windows.zip
-rw-r--r-- 1 root root 24335388 Apr 11 20:32 influxdb3-3.0.0-0.alpha.1_darwin_amd64.tar.gz
-rw-r--r-- 1 root root 28810063 Apr 11 20:32 influxdb3-3.0.0-0.alpha.1_linux_amd64.tar.gz
-rw-r--r-- 1 root root 26864086 Apr 11 20:32 influxdb3-3.0.0-0.alpha.1_linux_arm64.tar.gz
-rw-r--r-- 1 root root 26746749 Apr 11 20:32 influxdb3-3.0.0_0.alpha.1-1.aarch64.rpm
-rw-r--r-- 1 root root 28542862 Apr 11 20:32 influxdb3-3.0.0_0.alpha.1-1.x86_64.rpm
-rw-r--r-- 1 root root 28812320 Apr 11 20:32 influxdb3_3.0.0-0.alpha.1-1_amd64.deb
-rw-r--r-- 1 root root 26863794 Apr 11 20:32 influxdb3_3.0.0-0.alpha.1-1_arm64.de
```

I then verified the package versions outside the docker container:
```
$ dpkg -I influxdb3_3.0.0-0.alpha.1-1_amd64.deb 
 new Debian package, version 2.0.
 size 28812320 bytes: control archive=793 bytes.
     279 bytes,    11 lines      control              
     322 bytes,     5 lines      md5sums              
      12 bytes,     1 lines   *  postinst             #!/bin/bash
     434 bytes,    22 lines   *  preinst              #!/bin/bash
 Package: influxdb3
 Version: 3.0.0-0.alpha.1-1
 License: MIT/Apache-2.0
 Vendor: InfluxData
 Architecture: amd64
 Maintainer: support@influxdb.com
 Installed-Size: 85717
 Section: default
 Priority: optional
 Homepage: https://influxdata.com
 Description: Monolithic time-series database.
```

```
$ rpm -qip influxdb3-3.0.0_0.alpha.1-1.x86_64.rpm
Name        : influxdb3
Version     : 3.0.0_0.alpha.1
Release     : 1
Architecture: x86_64
Install Date: (not installed)
Group       : default
Size        : 87774723
License     : MIT/Apache-2.0
Signature   : (none)
Source RPM  : influxdb3-3.0.0_0.alpha.1-1.src.rpm
Build Date  : Thu 11 Apr 2024 04:32:07 PM EDT
Build Host  : 18795ca60e10
Relocations : / 
Packager    : support@influxdb.com
Vendor      : InfluxData
URL         : https://influxdata.com
Summary     : Monolithic time-series database.
Description :
Monolithic time-series database.
```
I'll test the remaining variants. I'll only add another comment if I encounter breakage. :slightly_smiling_face:

Comment (jdstrand) [reply .circleci/packages/config.yaml:6]: > $ dpkg -I influxdb3_3.0.0-0.alpha.1-1_amd64.deb 
> ...
> Package: influxdb3
> Version: 3.0.0-0.alpha.1-1

@bnpfeife - note that with the alpha, beta and rc builds that have a tag of form `v3.0.0-0.alpha.1`, etc, we don't need to tag `-1` on the end, that is what the `-0` part of `-0.alpha.1` is for. We only need to tack on `-1` if the tag is of form `^v[0-9]+.[0-9]+.[0-9]+$` (or doesn't contain 'alpha', 'beta' or 'rc', either works). It's ok that it is there from a package manager point of view, but it is extraneous.

Comment (bnpfeife) [reply .circleci/packages/config.yaml:6]: @mgattozzi @jdstrand 

Sorry about the confusion between `release` and `default`. I think that `release` is a misnomer and implies that there is something special about the package. The _only_ difference is which `value` is used for the package version. In retrospect, if I were to re-write this portion of `ci-packager`, the `version` section would look like this instead:
```yaml
version:
# Matches "release" versions (e.g v3.0.0)
- match: '^v[0-9]+.[0-9]+.[0-9]+$'
  value: '{{env.CIRCLE_TAG[1:]}}'
# Matches "alpha" versions (e.g v3.0.0-0.alpha)
- match: '^v[0-9]+.[0-9]+.[0-9]+-[0-9].alpha'
  value: '{{env.CIRCLE_TAG[1:]}}'
# Catch-all versions
- match: '.*'
  value: '3.0.0+snapshot-{{env.CIRCLE_SHA1[:8]}}'
```

Comment (jdstrand) [reply .circleci/packages/config.yaml:6]: > $ rpm -qip influxdb3-3.0.0_0.alpha.1-1.x86_64.rpm
> Name        : influxdb3
> Version     : 3.0.0_0.alpha.1

@bnpfeife - I expected the rpm build to use a `-` instead of an `_` here in both the file name and the internal version.

**Review by jdstrand — CHANGES_REQUESTED** [review]
Ok, with the additional conversation, I'd like to see these changes:
1. use `match: '^v[0-9]+.[0-9]+.[0-9]+(-[0-9]+\.(alpha|beta|rc)\.[0-9]+(\.[0-9]+)?)?$'`
2. don't append `-1` to deb builds for alpha, beta and rc
3. update the rpm versions to use `-` instead of `_` in the filename and internal version

**Comment by jdstrand** [comment]
> Hey @jdstrand, I've been looking into updating `ci-packager` to support these changes:
>
>> 2. don't append -1 to deb and rpm builds for alpha, beta and rc
>> 3. update the rpm versions to use - instead of _ in the filename and internal version
>
> I mentioned in out Slack that ci-packager executes fpm. fpm then modifies the version string before passing it to rpmbuild or dpkg-buildpackage. So, I experimented with calling rpmbuild and dpkg-buildpackage from ci-packager instead.
>
> Ultimately, I ended up consulting both the debian and redhat versioning guidelines quite a bit. Both packaging formats consider the hyphen a separator between the "upstream version" and "package version". These components have slightly different names in the actual version specifications. However, the purpose of each component remains largely the same.
>
> I _can_ definitely make hyphens work, but I'm not sure that I like the implications. I interpret the "release version" (`$CIRCLE_TAG`) to correspond entirely to the "upstream version". Splitting the "release version" into an "upstream version" and "package version" indicates that multiple releases of the package have occurred when upstream remained unchanged. 

Not having hyphens has packaging implications. In Debian, packages without hyphens are interpreted as Debian 'native' packages that come from Debian itself (eg, `dpkg` has version `1.21.1` in Debian whereas `postgresql-14` has `14.2-1`) and packaging is different between native and non-native packages. For releases of our software coming from git tags, we should use hyphens since we aren't a native package. One goal of the suggestion I gave is that `3.0.0-1` is intended to be 'the official release of 3.0.0' and '3.0.1-1' is "the official release of 3.0.1' whereas '3.0.1-2' is "the official release of 3.0.1 with a packaging change". This is fully inline with what you just said (but see below).

> It looks like `3.0.0-0.alpha.0` and the other pre-release versions have been crafted to sort _before_ `3.0.0-1`. I'm okay with this; ci-packager will just need to supply `1` as the "package version" when none could be parsed from the "release version".

Indeed.

> With all of that being said, I'm wondering if it would be easier just to use `~` and `+`? In both packaging formats, `~` sorts before other components. Outside of that, neither impart special meaning or split the "release version".
> 
> ```sh
> # debian version comparisons
> $ dpkg --compare-versions '1.0.0' gt '1.0.0~alpha0' # exit-code 0
> $ dpkg --compare-versions '1.0.0' gt '1.0.0+snapshot0' # exit-code 1
> 
> # redhat version comparisons
> $ rpmdev-vercmp '1.0.0' '1.0.0~alpha0'
> 1.0.0 > 1.0.0~alpha0
> $ rpmdev-vercmp '1.0.0' '1.0.0+snapshot0'
> 1.0.0 < 1.0.0+snapshot0
> ```

The problem is that in addition to deb and rpm packaging, I was trying to also reconcile [semver](https://semver.org/) versioning, and it doesn't allow a `~` or a `+` after the `x.y.z`, only `-`. I felt semver was important because this is coming from a git tag, git tags are shown in GitHub, more people are probably familiar with semver than esoteric Linux packaging version semantics and also we've seen Enterprise import `influxdb` in go.mod so I wanted good semver in case Community/Pro wanted to do something similar in Cargo.toml. I considered using `2.99.N` for our alpha and beta builds, but I wanted the `3.0.0` to make it extremely clear that we are talking about InfluxDB 3, not any version of 2.x. Therefore, I took a common approach and used `-0` to denote "this is unofficial" in the deb/rpm world with "this is prerelease" in the semver world, and then created the scheme I did that works for all of deb, rpm and semver for releases of our software that come from git tags:

> 1. use match: '^v[0-9]+.[0-9]+.[0-9]+(-[0-9]+\.(alpha|beta|rc)\.[0-9]+(\.[0-9]+)?)?$'
> 2. don't append -1 to deb and rpm builds for alpha, beta and rc (ed. since they already have `-0...`)
> 3. update the rpm versions to use - instead of _ in the filename and internal version

Yes, the alpha, beta and rc versions have rather verbose version, but it works from a packaging perspective and alpha, beta and rc builds are for internal use (eg, testing) so I wanted to also optimize for `MAJOR.MINOR.MICRO-N` versions of our software since by far most users will only ever use them. My goal for the alpha, beta and rc versioning is really just to be packaging friendly).

That leaves `+snapshot` where we agree: `3.0.0+snapshot-{{env.CIRCLE_SHA1[:8]}}`. In this, I abandoned the `-Nsomething` since a) these aren't coming from git tags so semver doesn't matter and b) I wanted the to convey the idea of 'this is 3.0.0 plus stuff' (so it is greater than 3.0.0-1 but less than 3.0.1-1).

In the end, the package versioning I chose is a compromise that:
* works with deb, rpm and semver package tooling with release versions coming from git tags which supports alpha, beta, rc and final release versions
* uses recommended policy for deb and rpm for official final versions (ie, 3.0.0-1, 3.0.1-1, 3.0.1-2, etc)
* uses established convention for deb and rpm for prerelease versions (ie, uses `-0something`)
* follows semver policy for prerelease versions (ie, uses `-0something`)
* use recommended policy for deb and rpm for unofficial snapshot releases that don't have a corresponding semver git tag (ie, 3.0.0+snapshot)

**Comment by mgattozzi** [comment]
@jdstrand and @bnpfeife I'm back from vacation and so trying to figure out what else needs to be done to get this over the line. Are there any actual changes to the versioning that need to be done here?

**Comment by jdstrand** [comment]
> @jdstrand and @bnpfeife I'm back from vacation and so trying to figure out what else needs to be done to get this over the line. Are there any actual changes to the versioning that need to be done here?

From my perspective, these are the changes:

>> 1. use match: '^v[0-9]+.[0-9]+.[0-9]+(-[0-9]+.(alpha|beta|rc).[0-9]+(.[0-9]+)?)?$'

config.yaml still has:
```
version:
  release:
    match: '^v[0-9]+.[0-9]+.[0-9]+'
    value: '{{env.CIRCLE_TAG[1:]}}'
  default:
    value: '3.0.0+snapshot-{{env.CIRCLE_SHA1[:8]}}'
```

This should be changed to:
```
version:
  release:
    match: '^v[0-9]+.[0-9]+.[0-9]+(-[0-9]+.(alpha|beta|rc).[0-9]+(.[0-9]+)?)?$'
    value: '{{env.CIRCLE_TAG[1:]}}'
  default:
    value: '3.0.0+snapshot-{{env.CIRCLE_SHA1[:8]}}'
```

>> 2. don't append -1 to deb and rpm builds for alpha, beta and rc (ed. since they already have -0...)
>> 3. update the rpm versions to use - instead of _ in the filename and internal version

I believe that @bnpfeife fixed these outside of this PR? @bnpfeife, can you comment?

**Comment by mgattozzi** [comment]
Okay I pushed those changes @jdstrand. If that looks good, I'll rebase off main and make sure CI passes, then add the correct filters so it won't run unless we want to do a release.

File: foo
Comment (jdstrand) [inline foo:1]: What's this file?

File: Dockerfile
Comment (jdstrand) [inline Dockerfile:60]: Should this be port 8086?

**Review by jdstrand — APPROVED** [review]
For the portions I reviewed before, LGTM. I noticed a new file `foo` that seems like a mistake and have a question about the listening port inline. In the interest of time, I'm going to approve, but please remove the `foo` file and consider my question about the port.

Please have @bnpfeife also give it a once over.

File: foo
Comment (mgattozzi) [reply foo:1]: Leftover from me testing stuff with sha256sum. I thought I had deleted it, but I think I did that in the web UI and forgot to pull in those changes before a push 🤦

**Comment by bnpfeife** [comment]
Hey, @jdstrand @mgattozzi 

I created a PR into this branch that uses a new version of `ci-packager`. It should support everything that was mentioned prior in this PR:
https://github.com/influxdata/influxdb/pull/24945

**Comment by mgattozzi** [comment]
@bnpfeife I've merged it here and if we need any more changes as follow up I can do them here, but I think this will get us over the edge.

**Review by bnpfeife — APPROVED** [review]

**Comment by mgattozzi** [comment]
Okay tomorrow I'm going to do some rebasing again, put the correct filters on, and then merge!

**Pull request merged**
